internal/tun/tun.go

```diff
@@ -0,0 +1,202 @@
+package tun
+
+// TODO(ainghazal): this package shares a bunch of code with tlsbio, consider
+// refactoring common parts.
+
+import (
+	"bytes"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/apex/log"
+	"github.com/ooni/minivpn/internal/model"
+	"github.com/ooni/minivpn/internal/networkio"
+	"github.com/ooni/minivpn/internal/session"
+)
+
+// StartTUN initializes and starts the TUN device over the vpn.
+func StartTUN(conn networkio.FramingConn, options *model.Options) *TUN {
+	// create a session
+	sessionManager, err := session.NewManager(log.Log)
+	if err != nil {
+		log.WithError(err).Fatal("tun.StartTUN")
+	}
+
+	// create the TUN that will OWN the connection
+	tunnel := newTUN(log.Log, conn, sessionManager)
+
+	// start all the workers
+	workers := startWorkers(log.Log, sessionManager, tunnel, conn, options)
+	tunnel.whenDone(func() {
+		workers.StartShutdown()
+		workers.WaitWorkersShutdown()
+	})
+
+	// Await for the signal from the session manager to known we're ready to start accepting data.
+	// In practice, this means that we already have a valid TunnelInfo at this point
+	// (i.e., three way handshake has completed, and we have valid keys).
+
+	// TODO(ainghazal): we need to timeout here.
+	<-sessionManager.Ready
```

Comment:
if you pass a context as the first argument, you can await either for session manager to be ready or for the context to be done, and then you have your timeout implementation, which I think we should add in this PR.